File: lib/model/timetable.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:tutility/model/semester.dart';
import 'package:tutility/model/subject.dart';

part 'timetable.freezed.dart';
part 'timetable.g.dart';

@freezed
class Timetable with _$Timetable {
  const factory Timetable({
    required int year,
    required String belong,
    required Semester semester,
    required List<List<Subject?>> firstHalf,
    required List<List<Subject?>> secondHalf,
    required List<List<Subject?>> intensive,
  }) = _Timetable;

  factory Timetable.fromJson(Map<String, dynamic> json) =>
      _$TimetableFromJson(json);
}

File: schemas/timetable.ts
import { z } from 'zod';
import { Subject } from './subject';

export const Timetable = z.object({
  year: z.string(),
  belong: z.string(),
  semester: z.enum(['spring', 'fall']),
  firstHalf: z.array(z.array(z.nullable(Subject))),
  secondHalf: z.array(z.array(z.nullable(Subject))),
  intensive: z.array(z.array(z.nullable(Subject))),
});

export type Timetable = z.infer<typeof Timetable>;
